pdi-execution-engine/impl/kettle-native/src/main/java/org/pentaho/di/engine/kettlenative/impl/Engine.java

```diff
@@ -0,0 +1,91 @@
+package org.pentaho.di.engine.kettlenative.impl;
+
+import org.pentaho.di.core.KettleEnvironment;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.engine.api.IExecutableOperation;
+import org.pentaho.di.engine.api.IEngine;
+import org.pentaho.di.engine.api.IOperation;
+import org.pentaho.di.engine.api.ITransformation;
+import org.pentaho.di.engine.api.ITuple;
+import rx.Observable;
+import rx.Subscriber;
+
+import java.util.AbstractMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class Engine implements IEngine {
+
+  @Override public void execute( ITransformation trans ) {
+    try {
+      KettleEnvironment.init();
+    } catch ( KettleException e ) {
+      e.printStackTrace();
+    }
+    List<IExecutableOperation> execOps = getExecutableOperations( trans );
+
+    wireExecution( execOps );
+
+    // start execution
+    execOps.stream()
+      .forEach( IExecutableOperation::start );
+
+    // wait till done
+    while( running( execOps ) ) {
+      try {
+        Thread.sleep( 50 );
+      } catch ( InterruptedException e ) {
+        e.printStackTrace();
+      }
+    }
+
+  }
+
+  private List<IExecutableOperation> getExecutableOperations( ITransformation trans ) {
+    return trans.getOperations()
+        .stream()
+        .map( KettleNativeExecutable::compile )
+        .collect( Collectors.toList() );
+  }
+
+  private boolean running( List<IExecutableOperation> execOps ) {
+    return execOps.stream()
+      .filter( IExecutableOperation::isRunning )
+      .findFirst()
+      .map( (op) -> new Boolean( true ) )
+      .orElse( false );
+  }
+
+  private void wireExecution( List<IExecutableOperation> execOps ) {
+    Map<IExecutableOperation, Observable<? super ITuple>> observables =
+      execOps.stream()
+        .collect( Collectors.toMap(
+          execOp -> execOp, execOp -> Observable.create( sub -> execOp.subscribe( sub ) ) ) );
+
+    observables.keySet().stream()
```

Comment:
Yeah, that block is pretty opaque and can definitely be made more intelligible.  The map of operation->observable is just used to make the subsequent creation of Subscribers for each "from" op simpler.